In [12]:
import pandas as pd

In [13]:
data = pd.read_csv("D:\STUDY\Coding\Jupyter Notebook\DATA SCIENCE\EDA\LOAN DATA analysis\DATA SET\Loan 1.csv")

#  Pandas DataFrame

- A Pandas DataFrame is a two-dimensional data structure used in Python for data analysis. 
- It is similar to a table or a spreadsheet, where data is organized in rows and columns. 
- Each column has a name, and the data within the column can be of different types such as numeric, categorical, or text.
- DataFrames are versatile and allow for a variety of data manipulation and analysis tasks, including filtering, sorting, grouping, merging, and visualizing.
- We can create it using **DataFrame()** function

## Creating a DataFrame

Here, we create a Pandas DataFrame with the given sample data to demonstrate its structure and basic operations.

In [14]:
df = pd.DataFrame(data)

In [15]:
# Display the DataFrame
print(df)

      Customer ID              Name Gender  Age  Income (USD)  \
0         C-26247  Tandra Olszewski      F   47       3472.69   
1         C-35067     Jeannette Cha      F   57       1184.84   
2         C-34590      Keva Godfrey      F   52       1266.27   
3         C-16668      Elva Sackett      M   65       1369.72   
4         C-12196    Sade Constable      F   60       1939.23   
...           ...               ...    ...  ...           ...   
19995      C-9076    Tobias Davilla      F   19       1349.60   
19996     C-17587    Evelina Hodges      M   22       2019.78   
19997     C-46479   Karlyn Mckinzie      M   19       2252.03   
19998      C-3099    Mariana Pulver      F   21       1845.35   
19999     C-43482   Mechelle Boddie      M   21       5233.16   

      Income Stability            Profession Type of Employment    Location  \
0                  Low  Commercial associate           Managers  Semi-Urban   
1                  Low               Working        Sales sta

In [16]:
# or simply
df

Customer ID              Name Gender  Age  Income (USD)  \
0         C-26247  Tandra Olszewski      F   47       3472.69   
1         C-35067     Jeannette Cha      F   57       1184.84   
2         C-34590      Keva Godfrey      F   52       1266.27   
3         C-16668      Elva Sackett      M   65       1369.72   
4         C-12196    Sade Constable      F   60       1939.23   
...           ...               ...    ...  ...           ...   
19995      C-9076    Tobias Davilla      F   19       1349.60   
19996     C-17587    Evelina Hodges      M   22       2019.78   
19997     C-46479   Karlyn Mckinzie      M   19       2252.03   
19998      C-3099    Mariana Pulver      F   21       1845.35   
19999     C-43482   Mechelle Boddie      M   21       5233.16   

      Income Stability            Profession Type of Employment    Location  \
0                  Low  Commercial associate           Managers  Semi-Urban   
1                  Low               Working        Sales staff       Rural   
2                  Low               Working                NaN  Semi-Urban   
3                 High             Pensioner                NaN       Rural   
4                 High             Pensioner                NaN       Urban   
...                ...                   ...                ...         ...   
19995              Low  Commercial associate                NaN  Semi-Urban   
19996              Low               Working         Core staff       Urban   
19997              Low               Working         Core staff  Semi-Urban   
19998              Low               Working                NaN  Semi-Urban   
19999              Low               Working           Laborers  Semi-Urban   

       Loan Amount Request (USD)  ...  Dependents Credit Score  \
0                      137088.98  ...         2.0       799.14   
1                      104771.59  ...         2.0       833.31   
2                      176684.91  ...         3.0       627.44   
3                       97009.18  ...         2.0       833.20   
4                      109980.00  ...         NaN          NaN   
...                          ...  ...         ...          ...   
19995                  156766.97  ...         4.0       684.32   
19996                   47924.80  ...         NaN       706.34   
19997                   18629.88  ...         1.0       656.46   
19998                   95430.73  ...         2.0       865.46   
19999                  107966.03  ...         3.0       857.16   

      No. of Defaults  Has Active Credit Card  Property ID  Property Age  \
0                   0             Unpossessed          843       3472.69   
1                   0             Unpossessed           22       1184.84   
2                   0             Unpossessed            1       1266.27   
3                   0                Inactive          730       1369.72   
4                   0                     NaN          356       1939.23   
...               ...                     ...          ...           ...   
19995               0                Inactive          681       1349.60   
19996               0                Inactive          213       2019.78   
19997               0                Inactive          270       2252.03   
19998               0             Unpossessed          489       1845.35   
19999               1                  Active          368       5233.16   

      Property Type  Property Location  Co-Applicant  Property Price  
0                 2              Urban             1        236644.5  
1                 1              Rural             1        142357.3  
2                 1              Urban             1       300991.24  
3                 1         Semi-Urban             0        125612.1  
4                 4         Semi-Urban             1          180908  
...             ...                ...           ...             ...  
19995             4         Semi-Urban             1          212778  
19

This code snippet creates a DataFrame with the provided data and displays it, giving you a sense of the structure and organization of the dataset.

### Creating a DataFrame from Dictionaries

- Pandas DataFrames are a convenient way to work with tabular data in Python. 
- You can create a DataFrame from various sources, such as dictionaries, lists, or CSV files. Let's start by creating a 
- DataFrame from a dictionary to demonstrate basic structure and operations.

In [17]:
# Create a dictionary with the provided data
data1 = {
    "Customer ID": ["C-26247", "C-35067", "C-34590", "C-16668", "C-12196", "C-2600", "C-9047", "C-2206", "C-25607", "C-11606"],
    "Name": ["Tandra Olszewski", "Jeannette Cha", "Keva Godfrey", "Elva Sackett", "Sade Constable", "Gina Weir", "Lacey Cybulski", "Karrie Dickison", 
             "Tona Wigington", "Hellen Alexis"],
    "Gender": ["F", "F", "F", "M", "F", "F", "M", "M", "M", "F"],
    "Age": [47, 57, 52, 65, 60, 59, 43, 65, 64, 27],
    "Income (USD)": [3472.69, 1184.84, 1266.27, 1369.72, 1939.23, 2944.81, 1957.31, 1403.63, 1604.65, 949.17],
    "Income Stability": ["Low", "Low", "Low", "High", "High", "Low", "Low", "High", "Low", "Working"],
    "Profession": ["Commercial associate", "Working", "Working", "Pensioner", "Pensioner", "Working", "Working", "Pensioner", "Working", "Working"],
    "Type of Employment": ["Managers", "Sales staff", "", "", "", "Sales staff", "Sales staff", "", "", "Laborers"],
    "Location": ["Semi-Urban", "Rural", "Semi-Urban", "Rural", "Urban", "Semi-Urban", "Rural", "Semi-Urban", "Semi-Urban", "Rural"],
    "Loan Amount Request (USD)": [137088.98, 104771.59, 176684.91, 97009.18, 109980, 31465.78, 150334.11, 121029.27, 39475.31, 24703.89]
}

In [18]:
#Creating DATAFRAME and printing it 
data2 = pd.DataFrame(data1)
data2

Customer ID              Name Gender  Age  Income (USD) Income Stability  \
0     C-26247  Tandra Olszewski      F   47       3472.69              Low   
1     C-35067     Jeannette Cha      F   57       1184.84              Low   
2     C-34590      Keva Godfrey      F   52       1266.27              Low   
3     C-16668      Elva Sackett      M   65       1369.72             High   
4     C-12196    Sade Constable      F   60       1939.23             High   
5      C-2600         Gina Weir      F   59       2944.81              Low   
6      C-9047    Lacey Cybulski      M   43       1957.31              Low   
7      C-2206   Karrie Dickison      M   65       1403.63             High   
8     C-25607    Tona Wigington      M   64       1604.65              Low   
9     C-11606     Hellen Alexis      F   27        949.17          Working   

             Profession Type of Employment    Location  \
0  Commercial associate           Managers  Semi-Urban   
1               Working        Sales staff       Rural   
2               Working                     Semi-Urban   
3             Pensioner                          Rural   
4             Pensioner                          Urban   
5               Working        Sales staff  Semi-Urban   
6               Working        Sales staff       Rural   
7             Pensioner                     Semi-Urban   
8               Working                     Semi-Urban   
9               Working           Laborers       Rural   

   Loan Amount Request (USD)  
0                  137088.98  
1                  104771.59  
2                  176684.91  
3                   97009.18  
4                  109980.00  
5                   31465.78  
6                  150334.11  
7                  121029.27  
8                   39475.31  
9                   24703.89

So, using Dataframe function we can create database from the dictionaries of values that we had.

### Creating a DataFrame from a CSV File

You can also create a DataFrame by reading data from a CSV file. This is useful when working with large datasets or when you need to select specific columns.

In [19]:
# Define the columns to select from the CSV file
selected_columns = ["Customer ID", "Name", "Gender", "Income (USD)"]

# Read the CSV file and select specific columns
df_from_csv = pd.read_csv("D:\STUDY\Coding\Jupyter Notebook\DATA SCIENCE\EDA\LOAN DATA analysis\DATA SET\Loan 1.csv", usecols=selected_columns)

# Display the first few rows to confirm the selection
df_from_csv.head()

Customer ID              Name Gender  Income (USD)
0     C-26247  Tandra Olszewski      F       3472.69
1     C-35067     Jeannette Cha      F       1184.84
2     C-34590      Keva Godfrey      F       1266.27
3     C-16668      Elva Sackett      M       1369.72
4     C-12196    Sade Constable      F       1939.23

---

# Data Inspection

- Data inspection is the process of examining a DataFrame to understand its structure, check for missing or inconsistent data, and learn about the data types of each column. 
- It provides insights into potential data issues and guides further data cleaning and analysis.

## Data Inspection Functions and their Meaning

1. **df.info()**: Provides information about the DataFrame, including the number of rows, columns, data types, and any missing data.
2. **df.head(n)**: Displays the first n rows of the DataFrame, useful for a quick overview.
3. **df.describe()**: Generates summary statistics for numerical columns, including count, mean, standard deviation, minimum, maximum, and quartiles.
4. **df.isnull().sum()**: Returns the count of missing values in each column, helping identify data quality issues.
5. **df.columns**: Lists the column names in the DataFrame.
6. **df.dtypes**: Shows the data types for each column, useful for detecting inconsistencies.

In [20]:
# Check the structure and basic information of the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 23 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Customer ID                  20000 non-null  object 
 1   Name                         20000 non-null  object 
 2   Gender                       19969 non-null  object 
 3   Age                          20000 non-null  int64  
 4   Income (USD)                 19250 non-null  float64
 5   Income Stability             19187 non-null  object 
 6   Profession                   20000 non-null  object 
 7   Type of Employment           15311 non-null  object 
 8   Location                     20000 non-null  object 
 9   Loan Amount Request (USD)    20000 non-null  float64
 10  Current Loan Expenses (USD)  19917 non-null  float64
 11  Expense Type 1               20000 non-null  object 
 12  Expense Type 2               20000 non-null  object 
 13  Dependents      

In [21]:
# Display the first 5 rows to inspect the data
df.head(5)

Customer ID              Name Gender  Age  Income (USD) Income Stability  \
0     C-26247  Tandra Olszewski      F   47       3472.69              Low   
1     C-35067     Jeannette Cha      F   57       1184.84              Low   
2     C-34590      Keva Godfrey      F   52       1266.27              Low   
3     C-16668      Elva Sackett      M   65       1369.72             High   
4     C-12196    Sade Constable      F   60       1939.23             High   

             Profession Type of Employment    Location  \
0  Commercial associate           Managers  Semi-Urban   
1               Working        Sales staff       Rural   
2               Working                NaN  Semi-Urban   
3             Pensioner                NaN       Rural   
4             Pensioner                NaN       Urban   

   Loan Amount Request (USD)  ...  Dependents Credit Score No. of Defaults  \
0                  137088.98  ...         2.0       799.14               0   
1                  104771.59  ...         2.0       833.31               0   
2                  176684.91  ...         3.0       627.44               0   
3                   97009.18  ...         2.0       833.20               0   
4                  109980.00  ...         NaN          NaN               0   

   Has Active Credit Card  Property ID  Property Age Property Type  \
0             Unpossessed          843       3472.69             2   
1             Unpossessed           22       1184.84             1   
2             Unpossessed            1       1266.27             1   
3                Inactive          730       1369.72             1   
4                     NaN          356       1939.23             4   

   Property Location  Co-Applicant  Property Price  
0              Urban             1        236644.5  
1              Rural             1        142357.3  
2              Urban             1       300991.24  
3         Semi-Urban             0        125612.1  
4         Semi-Urban             1          180908  

[5 rows x 23 columns]

In [22]:
# Generate summary statistics for numerical columns
df.describe()

Age  Income (USD)  Loan Amount Request (USD)  \
count  20000.000000  19250.000000               20000.000000   
mean      39.940500   2536.995394               88859.076706   
std       16.097618   1415.908803               60007.333673   
min       18.000000    368.590000                6185.480000   
25%       25.000000   1659.722500               40889.030000   
50%       40.000000   2224.590000               74565.990000   
75%       55.000000   3082.925000              120068.852500   
max       65.000000  30427.680000              576335.680000   

       Current Loan Expenses (USD)    Dependents  Credit Score  \
count                 19917.000000  18858.000000  19257.000000   
mean                    408.462359      2.251246    738.821538   
std                     215.846532      0.953153     72.406504   
min                      41.370000      1.000000    580.040000   
25%                     250.890000      2.000000    680.300000   
50%                     374.000000      2.000000    739.300000   
75%                     519.570000      3.000000    797.440000   
max                    3305.970000     13.000000    893.830000   

       No. of Defaults   Property ID  Property Age  Property Type  
count     20000.000000  20000.000000  19108.000000    20000.00000  
mean          0.189500    501.024500   2534.102003        2.47190  
std           0.391915    287.537722   1408.509255        1.12551  
min           0.000000      1.000000    368.590000        1.00000  
25%           0.000000    252.000000   1658.397500        1.00000  
50%           0.000000    501.000000   2220.605000        2.00000  
75%           0.000000    748.000000   3082.250000        3.00000  
max           1.000000    999.000000  30427.680000        4.00000

In [23]:
# Check for missing values
print("Missing values per column:")
df.isnull().sum()

Missing values per column:


Customer ID                       0
Name                              0
Gender                           31
Age                               0
Income (USD)                    750
Income Stability                813
Profession                        0
Type of Employment             4689
Location                          0
Loan Amount Request (USD)         0
Current Loan Expenses (USD)      83
Expense Type 1                    0
Expense Type 2                    0
Dependents                     1142
Credit Score                    743
No. of Defaults                   0
Has Active Credit Card         1076
Property ID                       0
Property Age                    892
Property Type                     0
Property Location               160
Co-Applicant                      0
Property Price                    0
dtype: int64

In [24]:
#Lists the column names in the DataFrame.
df.columns

Index(['Customer ID', 'Name', 'Gender', 'Age', 'Income (USD)',
       'Income Stability', 'Profession', 'Type of Employment', 'Location',
       'Loan Amount Request (USD)', 'Current Loan Expenses (USD)',
       'Expense Type 1', 'Expense Type 2', 'Dependents', 'Credit Score',
       'No. of Defaults', 'Has Active Credit Card', 'Property ID',
       'Property Age', 'Property Type', 'Property Location', 'Co-Applicant',
       'Property Price'],
      dtype='object')

In [25]:
#Shows the data types for each column
df.dtypes

Customer ID                     object
Name                            object
Gender                          object
Age                              int64
Income (USD)                   float64
Income Stability                object
Profession                      object
Type of Employment              object
Location                        object
Loan Amount Request (USD)      float64
Current Loan Expenses (USD)    float64
Expense Type 1                  object
Expense Type 2                  object
Dependents                     float64
Credit Score                   float64
No. of Defaults                  int64
Has Active Credit Card          object
Property ID                      int64
Property Age                   float64
Property Type                    int64
Property Location               object
Co-Applicant                    object
Property Price                  object
dtype: object

---

# Data Exploration 

Data exploration is the process of analyzing the DataFrame to understand distributions, patterns, and relationships within the data. It helps to generate hypotheses and guide further analysis or data manipulation.

Data Exploration Functions and their Meaning
- **df["column"].value_counts()**: Returns the count of unique values in a categorical column, helping understand the distribution.
- **df["column"].unique()**: Lists all unique values in a column, useful for understanding diversity in the data.
- **df.corr()**: Computes the correlation matrix for numerical columns, indicating relationships between variables.

There are other techniques too which involves plots like histograms, heatmaps and scatterplot which we will learn and discuss in seaborn and matplotlib.

### .value_counts()

In [26]:
# Get counts of unique values in the 'Gender' column
gender_counts = df["Gender"].value_counts()
print("Gender Counts:")
print(gender_counts)

Gender Counts:
F    10142
M     9827
Name: Gender, dtype: int64


Shows there are 10142 Females whereas 9827 males in the 'gender' columns

In [27]:
# Get the count of unique values in the "Location" column
location_counts = df["Location"].value_counts()
print("Location Counts:")
print(location_counts)

Location Counts:
Semi-Urban    14463
Rural          3436
Urban          2101
Name: Location, dtype: int64


Shows the number of location counts

### .unique()

In [28]:
# Get the unique values in the "Gender" column
unique_genders = df["Gender"].unique()
print("Unique Genders:")
print(unique_genders)

Unique Genders:
['F' 'M' nan]


Shows the number of unique values in the gender column

In [29]:
# Get the unique values in the "Location" column
unique_locations = df["Location"].unique()
print("Unique Locations:")
print(unique_locations)

Unique Locations:
['Semi-Urban' 'Rural' 'Urban']


Shows the number of unique values in the 'location' column

### corr()

Correlation refers to a statistical measure that describes the degree to which two or more variables move in relation to each other.
- A correlation coefficien, can range from -1 to +1

Here, it is shown in the form of *Correlation Matrix*

In [30]:
correlation_matrix = data2.corr()

print("Correlation Matrix:")
print(correlation_matrix)

Correlation Matrix:
                                Age  Income (USD)  Loan Amount Request (USD)
Age                        1.000000      0.048481                   0.068921
Income (USD)               0.048481      1.000000                   0.047194
Loan Amount Request (USD)  0.068921      0.047194                   1.000000


C:\Users\manis\AppData\Local\Temp\ipykernel_16888\3590295665.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  correlation_matrix = data2.corr()
